File: NyaLauncher.Core/Config/GlobalLaunchSettings.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using NyaLauncher.Core.Config;

namespace NyaLauncher.Core.Config;

public sealed record GlobalLaunchSettings(
    int WindowWidth,
    int WindowHeight,
    string JavaExecutable,
    string[] AdditionalJvmArguments,
    string[] AdditionalGameArguments);

/// <summary>
/// 持久化并加载全局高级启动设置（窗口大小、Java 路径、JVM/游戏参数）。
/// </summary>
public static class GlobalLaunchSettingsStore
{
    private const string WindowWidthKey = "globalLaunchWindowWidth";
    private const string WindowHeightKey = "globalLaunchWindowHeight";
    private const string JavaExecutableKey = "globalLaunchJavaExecutable";
    private const string JvmArgumentsKey = "globalLaunchJvmArguments";
    private const string GameArgumentsKey = "globalLaunchGameArguments";
    private const string AutomaticJavaValue = "$auto";

    public static GlobalLaunchSettings Load() => new(
        ReadInt(WindowWidthKey, 854, 320),
        ReadInt(WindowHeightKey, 480, 240),
        ReadJavaExecutable(),
        ReadArguments(JvmArgumentsKey),
        ReadArguments(GameArgumentsKey));

    public static bool Save(GlobalLaunchSettings settings)
    {
        ArgumentNullException.ThrowIfNull(settings);
        if (settings.WindowWidth < 320 || settings.WindowHeight < 240)
            return false;

        var normalized = settings with
        {
            JavaExecutable = (settings.JavaExecutable ?? string.Empty).Trim(),
            AdditionalJvmArguments = NormalizeArguments(settings.AdditionalJvmArguments),
            AdditionalGameArguments = NormalizeArguments(settings.AdditionalGameArguments)
        };
        return LauncherConfig.SetValue(WindowWidthKey, normalized.WindowWidth.ToString()) &&
               LauncherConfig.SetValue(WindowHeightKey, normalized.WindowHeight.ToString()) &&
               LauncherConfig.SetValue(
                   JavaExecutableKey,
                   string.IsNullOrWhiteSpace(normalized.JavaExecutable)
                       ? AutomaticJavaValue
                       : normalized.JavaExecutable) &&
               LauncherConfig.SetValue(
                   JvmArgumentsKey,
                   JsonSerializer.Serialize(normalized.AdditionalJvmArguments)) &&
               LauncherConfig.SetValue(
                   GameArgumentsKey,
                   JsonSerializer.Serialize(normalized.AdditionalGameArguments));
    }

    /// <summary>
    /// 仅保存窗口尺寸（宽/高），不触碰 Java 路径与 JVM/游戏参数。
    /// 用于主窗口尺寸变更时轻量落盘，避免整份设置回写，也避免用旧值覆盖其它字段。
    /// </summary>
    public static bool SaveWindowSize(int width, int height)
    {
        if (width < 320 || height < 240)
            return false;

        return LauncherConfig.SetValue(WindowWidthKey, width.ToString()) &&
               LauncherConfig.SetValue(WindowHeightKey, height.ToString());
    }

    private static int ReadInt(string key, int fallback, int minimum) =>
        int.TryParse(LauncherConfig.GetValue(key), out var value) && value >= minimum
            ? value
            : fallback;

    private static string ReadJavaExecutable()
    {
        var configured = LauncherConfig.GetValue(JavaExecutableKey);
        if (string.Equals(configured, AutomaticJavaValue, StringComparison.Ordinal))
            return string.Empty;
        return configured ?? LauncherConfig.JavaExecutable ?? string.Empty;
    }

    private static string[] ReadArguments(string key)
    {
        var value = LauncherConfig.GetValue(key);
        if (string.IsNullOrWhiteSpace(value))
            return [];
        try
        {
            return NormalizeArguments(JsonSerializer.Deserialize<string[]>(value));
        }
        catch (JsonException)
        {
            return [];
        }
    }

    private static string[] NormalizeArguments(IEnumerable<string>? arguments) =>
        arguments?
            .Where(argument => !string.IsNullOrWhiteSpace(argument))
            .Select(argument => argument.Trim())
            .ToArray() ?? [];
}

File: internal/config/globallaunchsettings.go
package config

import (
	"encoding/json"
	"strings"
)

// GlobalLaunchSettings 全局高级启动设置的内存快照。
type GlobalLaunchSettings struct {
	WindowWidth             int
	WindowHeight            int
	JavaExecutable          string
	AdditionalJvmArguments  []string
	AdditionalGameArguments []string
}

// GlobalLaunchSettingsStore 的配置键：所有字段写入 config.json 的独立键，
// 损坏的 JSON 值回退到默认值。
const (
	windowWidthKey    = "globalLaunchWindowWidth"
	windowHeightKey   = "globalLaunchWindowHeight"
	javaExecutableKey = "globalLaunchJavaExecutable"
	jvmArgumentsKey   = "globalLaunchJvmArguments"
	gameArgumentsKey  = "globalLaunchGameArguments"

	// automaticJavaValue Java 路径的"自动选择"占位值
	// （空路径在保存时写成它，避免与未配置混淆）。
	automaticJavaValue = "$auto"

	minimumWindowWidth  = 320
	minimumWindowHeight = 240
	defaultWindowWidth  = 854
	defaultWindowHeight = 480
)

// LoadGlobalLaunchSettings 加载全局高级启动设置。
func LoadGlobalLaunchSettings() GlobalLaunchSettings {
	return GlobalLaunchSettings{
		WindowWidth:             readInt(windowWidthKey, defaultWindowWidth, minimumWindowWidth),
		WindowHeight:            readInt(windowHeightKey, defaultWindowHeight, minimumWindowHeight),
		JavaExecutable:          readJavaExecutable(),
		AdditionalJvmArguments:  readArguments(jvmArgumentsKey),
		AdditionalGameArguments: readArguments(gameArgumentsKey),
	}
}

// SaveGlobalLaunchSettings 保存整份全局设置（参数先规范化：去空白、剔除空参数）。
// 尺寸过小时拒绝保存。
func SaveGlobalLaunchSettings(settings GlobalLaunchSettings) bool {
	if settings.WindowWidth < minimumWindowWidth ||
		settings.WindowHeight < minimumWindowHeight {
		return false
	}

	normalized := settings
	normalized.JavaExecutable = strings.TrimSpace(settings.JavaExecutable)
	normalized.AdditionalJvmArguments = normalizeArguments(settings.AdditionalJvmArguments)
	normalized.AdditionalGameArguments = normalizeArguments(settings.AdditionalGameArguments)
	// 五个键一次事务落盘，避免中途失败留下半新半旧的全局设置
	return UpdateInTransaction(func(config map[string]any) bool {
		config[windowWidthKey] = itoa(normalized.WindowWidth)
		config[windowHeightKey] = itoa(normalized.WindowHeight)
		if strings.TrimSpace(normalized.JavaExecutable) == "" {
			config[javaExecutableKey] = automaticJavaValue
		} else {
			config[javaExecutableKey] = normalized.JavaExecutable
		}
		config[jvmArgumentsKey] = serializeStringList(normalized.AdditionalJvmArguments)
		config[gameArgumentsKey] = serializeStringList(normalized.AdditionalGameArguments)
		return true
	})
}

// SaveGlobalWindowSize 仅保存窗口尺寸（宽/高），不触碰 Java 路径与 JVM/游戏参数。
// 用于主窗口尺寸变更时轻量落盘，避免整份设置回写，也避免用旧值覆盖其它字段。
func SaveGlobalWindowSize(width, height int) bool {
	if width < minimumWindowWidth || height < minimumWindowHeight {
		return false
	}
	return UpdateInTransaction(func(config map[string]any) bool {
		config[windowWidthKey] = itoa(width)
		config[windowHeightKey] = itoa(height)
		return true
	})
}

func readInt(key string, fallback, minimum int) int {
	if value, err := parseInt(GetValue(key)); err == nil && value >= minimum {
		return value
	}
	return fallback
}

func readJavaExecutable() string {
	configured := GetValue(javaExecutableKey)
	if configured == automaticJavaValue {
		return ""
	}
	// 从未配置过时回退到全局探测到的 Java
	if configured == "" {
		return JavaExecutable()
	}
	return configured
}

func readArguments(key string) []string {
	value := GetValue(key)
	if strings.TrimSpace(value) == "" {
		return []string{}
	}
	var arguments []string
	if err := json.Unmarshal([]byte(value), &arguments); err != nil {
		// 用户手改 config.json 写坏数组 → 回退为空参数，不影响启动
		return []string{}
	}
	return normalizeArguments(arguments)
}
